### PR TITLE
feature/EODHP-16-deploy-jupyter-notebook-to-cluster-3

### PR DESCRIPTION
**Pull request by sgillies-tpzuk**
Final tweak to get apphub to work. I overrode the EOEPCA pyconfig file to update the hardcoded kubespawner label filter with one that could be applied via Terraform to the node groups.

I also added cluster role and role bindings in addition to the existing role and role binding to allow RBAC permission to create new namespaces and required resources within those namespaces. This is required for user workspaces as spearate namespaces in Kubernetes Jupyter Hub (as indicated here https://jupyterhub-kubespawner.readthedocs.io/en/latest/spawner.html#kubespawner.KubeSpawner.enable_user_namespaces).

I had to use a generator rather than Kustomize configMapGenerator due to order of operations. configMapGenerator merge was attempted before Helm chart inflation, so required confgiMap did not yet exist otherwise. Kustomize generators are run after helm chart inflation.

I also took this opportunity to customise the apps in EODHP app-hub. I removed everything except JupyterLab and renamed is as "Workflow Analysis (JupyterLab).

To test, reviewers can visit https://dev.eodatahub.org.uk/apphub and log in as bob. Ask me for credentials. You should be able to start a server and select Workflow Analysis. Start a session and you should get a Jupyter Notebook.

### CODE REVIEW

File: apps/app-hub/envs/dev/files/jupyter_config.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    profile_slug = spawner.user_options.get("profile", None)
+
+    env = os.environ["JUPYTERHUB_ENV"].lower()
+
+    spawner.environment["CALRISSIAN_POD_NAME"] = f"jupyter-{spawner.user.name}-{env}"
+
+    spawner.log.info(f"Using profile slug {profile_slug}")
+
+    # namespace = f"jupyter-{spawner.user.name}"
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/jupyter_config.py:51]: Is it worth removing commented out code?

File: apps/app-hub/envs/dev/files/jupyter_config.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            "oauth_user": user_data_response,
+            "scope": scope,
+            "id_token": id_token,
+        }
+
+    async def pre_spawn_start(self, user, spawner):
+        """Pass upstream_token to spawner via environment variable"""
+        auth_state = await user.get_auth_state()
+        if not auth_state:
+            # auth_state not enabled
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/jupyter_config.py:149]: Could this comment be removed? The line above seems fairly self-explanatory

File: apps/app-hub/envs/dev/files/jupyter_config.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+c.KubeSpawner.namespace = app_hub_namespace
+
+# User namespace
+c.KubeSpawner.enable_user_namespaces = True
+c.KubeSpawner.user_namespace_template = (
+    resource_manager_workspace_prefix + "-{username}"
+)
+
+c.KubeSpawner.options_form = custom_options_form
+# c.KubeSpawner.image_pull_policy = "IfNotPresent"
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/jupyter_config.py:233]: Is it worth removing this line?

File: apps/app-hub/envs/dev/files/config.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        storage_class: block-storage
+        access_modes:
+          - "ReadWriteOnce"
+        volume_mount:
+          name: volume-workspace
+          mount_path: "/workspace"
+        persist: true
+    pod_env_vars:
+      A: "10"
+      B: "20"
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/config.yml:51]: Are these used or could/should they be removed?

File: apps/app-hub/envs/dev/files/jupyter_config.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+# rather than the pure-python implementations. The default one starts
+# being too slow to make a large number of requests to the proxy API
+# at the rate required.
+AsyncHTTPClient.configure("tornado.curl_httpclient.CurlAsyncHTTPClient")
+
+c.ConfigurableHTTPProxy.api_url = (
+    f'http://{get_name("proxy-api")}:{get_name_env("proxy-api", "_SERVICE_PORT")}'
+)
+
+print(f"c.ConfigurableHTTPProxy.api_url = {c.ConfigurableHTTPProxy.api_url}")
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/jupyter_config.py:93]: Is this still in use?

File: apps/app-hub/envs/dev/files/jupyter_config.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+# Don't wait at all before redirecting a spawning user to the progress page
+c.JupyterHub.tornado_settings = {
+    "slow_spawn_timeout": 0,
+}
+
+
+class EoepcaOAuthenticator(GenericOAuthenticator):
+    login_service = Unicode("EOEPCA", config=True)
+    id_token = None
+    print("LOGIN EOEPCA")
```
Comment (hcollingwood) [inline apps/app-hub/envs/dev/files/jupyter_config.py:105]: Is this still in use?

File: apps/app-hub/envs/dev/files/config.yml (hunk 4 above)
Comment (sgillies-tpzuk) [reply apps/app-hub/envs/dev/files/config.yml:51]: I don't know what they do to be honest, so i'm not comfortable removing them. There are no obvious docs on their usage.

File: apps/app-hub/envs/dev/files/jupyter_config.py (hunk 5 above)
Comment (sgillies-tpzuk) [reply apps/app-hub/envs/dev/files/jupyter_config.py:93]: I think it's for the purposes of the logs, i think I'll leave them.

File: apps/app-hub/envs/dev/files/jupyter_config.py (hunk 6 above)
Comment (sgillies-tpzuk) [reply apps/app-hub/envs/dev/files/jupyter_config.py:105]: I think it is, but you havve drawn my attention to the fact I can now rename the button from EOEPCA to EO DataHub. At least I think that's what this does.

File: apps/app-hub/envs/dev/files/config.yml (hunk 4 above)
Comment (hcollingwood) [reply apps/app-hub/envs/dev/files/config.yml:51]: They seem like example environment variables with some basic values defined but happy to leave them in

**Review by hcollingwood — APPROVED** [review]

**Pull request merged**